### PR TITLE
[WIP] Introduce Engine trait

### PR DESCRIPTION
**Pull request by 5kbpers**
## What have you changed? (mandatory)

This PR introduces a engine trait, which can replace the usages of rocksdb in store/raftstore/fsm/apply.rs

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

https://github.com/tikv/tikv/issues/4184

### CODE REVIEW

**Comment by brson** [comment]
@5kbpers I've finally started giving this a look. I pinged you with some questions on slack and am poking at it locally. I'll get back with some feedback soon.

**Comment by brson** [comment]
Hi @5kbpers @aknuds1.

This is starting to look really great. I see some of the engine abstractions are starting to get integrated, and that's encouraging.

I wonder @5kbpers can you try to do a couple of things:

- share your engine trait design publicly and paste a link here for @aknuds1
- get this PR building and passing CI and land what you've already got, so we can regroup and figure out the next step, how the three of us can coordinate

I don't know how to say this best, but I've been apprehensive about this PR, and am not sure how best I can help. This is a huge refactor and huge refactors are difficult.

The two things that worry me here are:

- the commits are big and change a lot at once
- there's a lot of code moving all around the codebase

The big commits make it difficult to review, to understand incrementally what the goals are and the steps toward the goals. The large amount of code motion makes the PR likely difficult to merge - other PRs will cause merge conflicts.

I don't know if this is helpful, but when I do a big refactor I like to follow a few principles:

- Make very small commits that perform one single action, after which the project still builds and tests
- Make small, self-contained PRs that make a coherent change to a single subsystem in service of the larger refactor

If I were to do this engine abstraction myself I would probably go about it the following way:

- Create an engine_traits crate and fill it in with some stub traits with no methods. Putting the traits in their own crate from the beginning establishes the abstraction boundary that can't be violated.
- Identify the simplest piece of code in tikv that uses an engine and fill in enough of the engine trait and impl to abstract that code only.
- Do the same over and over, in tiny steps.

Then, after succeeding at abstracting the engine once, throw all that work away as a useful prototype, and redo the sequence into a series of small but sensible PRs that con be easily evaluated and merged.

That's kind of why I think it would be good to wrap up _this_ PR and get it merged, step back and figure out what needs to happen next.

FWIW, for good background on refactoring techniques I recommend the books "Working effectively with legacy code", and "Refactoring: Improving the design of existing code".

This is just a quick braindump, but I'll follow up soon and try to be more constructive.

**Comment by 5kbpers** [comment]
@brson OK, this PR can pass CI now, but still has much work to do. These traits do not completely cover all usages of RocksDB in TiKV currently.
I will try to create an engine_traits crate, split this PR into some more smaller PRs, and land them.
And I will share a doc about current engine trait design and the usages of RocksDB in TiKV a little later. @aknuds1

**Comment by siddontang** [comment]
@aknuds1

I very want you to work us together to finish this big job. @brson can mentor you if possible.

**Comment by aknuds1** [comment]
Thanks @5kbpers!

@brson Is there anything for me to do on this in the meantime or shall I wait until this PR is landed?

File: temp [hunk 1]
```diff
@@ -0,0 +1,90 @@
+-                let snapshot = engine::Snapshot::new(Arc::clone(&self.db));
```
Comment (zhangjinpeng1987) [inline temp]: Please remove this file.

**Pull request closed without merging**